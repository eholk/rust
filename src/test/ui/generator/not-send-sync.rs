#![feature(generators)]

use std::cell::Cell;

fn main() {
    fn assert_sync<T: Sync>(_: T) {}
    fn assert_send<T: Send>(_: T) {}

    assert_sync(|| {
        //~^ ERROR: generator cannot be shared between threads safely
        let a = Cell::new(2);
        yield;
        // Explicity drop a so it is live across the yield for liveness-based generator captures.
        drop(a);
    });

    let a = Cell::new(2);
    assert_send(|| {
        //~^ ERROR: E0277
        drop(&a);
        yield;
    });
}
