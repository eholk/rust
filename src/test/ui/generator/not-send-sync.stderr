error[E0277]: `Cell<i32>` cannot be shared between threads safely
  --> $DIR/not-send-sync.rs:18:5
   |
LL |     assert_send(|| {
   |     ^^^^^^^^^^^ `Cell<i32>` cannot be shared between threads safely
   |
   = help: the trait `Sync` is not implemented for `Cell<i32>`
   = note: required because of the requirements on the impl of `Send` for `&Cell<i32>`
   = note: required because it appears within the type `[generator@$DIR/not-send-sync.rs:18:17: 22:6]`
note: required by a bound in `assert_send`
  --> $DIR/not-send-sync.rs:7:23
   |
LL |     fn assert_send<T: Send>(_: T) {}
   |                       ^^^^ required by this bound in `assert_send`

error: generator cannot be shared between threads safely
  --> $DIR/not-send-sync.rs:9:5
   |
LL |     assert_sync(|| {
   |     ^^^^^^^^^^^ generator is not `Sync`
   |
   = help: within `[generator@$DIR/not-send-sync.rs:9:17: 15:6]`, the trait `Sync` is not implemented for `Cell<i32>`
note: generator is not `Sync` as this value is used across a yield
  --> $DIR/not-send-sync.rs:11:13
   |
LL |         let a = Cell::new(2);
   |             ^
   |             |
   |             yield occurs here, with `a` maybe used later
   |             has type `Cell<i32>` which is not `Sync`
note: required by a bound in `assert_sync`
  --> $DIR/not-send-sync.rs:6:23
   |
LL |     fn assert_sync<T: Sync>(_: T) {}
   |                       ^^^^ required by this bound in `assert_sync`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0277`.
